ui/src/controller/track_decider.ts

```diff
@@ -2579,12 +2579,22 @@ class TrackDecider {
       }));
     }
 
+    const topLvlTracks = this.tracksToAdd.filter((track, index)=> {
+      if (!track.trackGroup || track.trackGroup === SCROLLING_TRACK_GROUP) {
+        this.tracksToAdd.splice(index, 1);
```

Comment:
We're splicing elements out of the array that we're filtering, which reads like it's broken. The `filter` method creates a new array, so that should be good enough.

It would be much more obvious to me (and less prone to future bugs when this code is changed) if what we did was to iterate the `tracksToAdd` twice:

- once filtering for the top level tracks, to push an action to add them
- again, this time filtering for the other tracks, to push an action to add those

Reply:
What if I did a foreach instead of the filter and filled an array instead using splice
```
const topLvlTracks  = [];
foreach((thing,index)=>{
       if (!track.trackGroup || track.trackGroup === SCROLLING_TRACK_GROUP) {
        topLvlTracks .push(this.tracksToAdd.splice(index, 1));
})
```
The idea is I want remove these tracks from the trackToAdd and put them into a seperate array.  If I were to do two seperate loops I would be looping over the same data twice to check once for if and another for else.

Reply:
Looping twice is still O(n) so not a problem for me. I just found it hard to understand and it looks like a bug, in part because in some languages it actually wouldn't even work. A foreach makes no difference but I won't argue the point.